src/control/pbin/exec.go

```diff
@@ -0,0 +1,148 @@
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+package pbin
+
+import (
+	"context"
+	"encoding/json"
+	"os/exec"
+
+	"github.com/pkg/errors"
+
+	"github.com/daos-stack/daos/src/control/logging"
+)
+
+const MaxMessageSize = 4096
+
+type (
+	// Request represents a request sent to the privileged binary. The
+	// payload field contains a JSON-encoded representation of the wrapped
+	// request.
+	Request struct {
+		Method  string
+		Payload json.RawMessage
+	}
+
+	// RequestFailure represents a failed request. The error message
+	// (if available) is stored as a string in the Message field.
+	RequestFailure struct {
+		Message string
+	}
+
+	// Response represents a response received from the privileged binary. The
+	// payload field contains a JSON-encoded representation of the wrapped response.
+	Response struct {
+		Error   *RequestFailure
+		Payload json.RawMessage
+	}
+
+	cmdLogger struct {
+		logFn  func(string)
+		prefix string
+	}
+)
+
+func (rf *RequestFailure) Error() string {
+	if rf == nil {
+		return "nil *RequestFailure"
+	}
+	return rf.Message
+}
+
+func (cl *cmdLogger) Write(data []byte) (int, error) {
+	if cl.logFn == nil {
+		return 0, errors.New("no log function set in cmdLogger")
+	}
+
+	var msg string
+	if cl.prefix != "" {
+		msg = cl.prefix + " "
+	}
+	msg += string(data)
+	cl.logFn(msg)
+	return len(data), nil
+}
+
+func ExecReq(parent context.Context, log logging.Logger, binPath string, req *Request) (res *Response, err error) {
+	if req == nil {
+		return nil, errors.New("nil request")
+	}
+
+	ctx, killChild := context.WithCancel(parent)
+	defer killChild()
+
+	child := exec.CommandContext(ctx, binPath)
+	child.Stderr = &cmdLogger{
+		logFn:  log.Error,
+		prefix: binPath,
+	}
+	toChild, err := child.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+	fromChild, err := child.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	conn := NewStdioConn("server", binPath, fromChild, toChild)
+
+	if err := child.Start(); err != nil {
+		return nil, err
+	}
+	// make sure we reap any children on the way out
+	defer func() {
+		// If there was an error, kill the child so that it can't
+		// hang around waiting for input.
+		if err != nil {
+			killChild()
+			return
+		}
+
+		// Otherwise, the child should exit normally.
+		err = child.Wait()
+	}()
+
+	sendData, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+	if _, err := conn.Write(sendData); err != nil {
+		return nil, errors.Wrap(err, "pbin write failed")
+	}
+
+	recvData := make([]byte, MaxMessageSize)
```

Comment:
as above comment regarding whether Max message size should be the capacity rather than total zeroed length